### PR TITLE
CBL token

### PR DESCRIPTION
**Pull request by Krishnakumarskr**

### CODE REVIEW

**Review by jplodge-pro — APPROVED** [review]
Looks good to me.

File: packages/contracts/resource/ci.toml [hunk 1]
```diff
@@ -19,6 +19,10 @@ custodian = "0x8561845F6a9511cD8e2daCae77A961e718A77cF6"
 # 2 decimal place percentage (meaining value divided by 100) as integer.
 collateral_percentage = 200
 
+[evm.contracts.token]
```
[the remark this answers is not in the document]
Comment (jplodge-pro) [reply packages/contracts/resource/ci.toml]: Krishna added `cbl`.

File: packages/contracts/script/HelperConfig.s.sol [hunk 2]
```diff
@@ -94,4 +103,13 @@ contract HelperConfig is Script {
 
         return factoryParams;
     }
+
+    function createTokenParamsFromConfig() internal view returns (TokenParams memory) {
+        TokenParams memory tokenParams = TokenParams({
+            owner: tomlConfig.readAddress(".evm.address.owner"),
+            maxSupply: tomlConfig.readUint(".evm.contracts.token.max_supply") * 1e18
```
[the remark this answers is not in the document]
Comment (jplodge-pro) [reply packages/contracts/script/HelperConfig.s.sol]: @Krishnakumarskr changed the configuration to be a string Wei value.

And I must admit, I am MUCH more inclined to revert it to the `10_000_000` value and formalise the conversion into a Wei value. Will discuss.

File: packages/contracts/src/token/CBL.sol [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+pragma solidity ^0.8.19;
+
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ERC20Permit } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
+import { ERC20Burnable } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
+import { Pausable } from "@openzeppelin/contracts/utils/Pausable.sol";
+
+contract CBL is ERC20, Ownable, ERC20Permit, ERC20Burnable, Pausable {
```
Comment (pasviegas) [inline packages/contracts/src/token/CBL.sol:11]: are we planing to have Ownable or AccessControl?

File: packages/contracts/src/token/CBL.sol [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    error CBL__MaxSupplyExceeded();
+
+    uint256 public maxSupply;
+
+    constructor(address _owner, uint256 _maxSupply) ERC20("Credbull", "CBL") ERC20Permit("test") Ownable(_owner) {
+        maxSupply = _maxSupply;
+    }
+
+    function mint(address to, uint256 amount) external onlyOwner {
+        if (totalSupply() + amount > maxSupply) {
```
Comment (pasviegas) [inline packages/contracts/src/token/CBL.sol:21]: what happens if we burn tokens? does the max supply goes down? if so, could we increase it again?

File: packages/contracts/src/token/CBL.sol [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import { ERC20Permit } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
+import { ERC20Burnable } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
+import { Pausable } from "@openzeppelin/contracts/utils/Pausable.sol";
+
+contract CBL is ERC20, Ownable, ERC20Permit, ERC20Burnable, Pausable {
+    error CBL__MaxSupplyExceeded();
+
+    uint256 public maxSupply;
+
+    constructor(address _owner, uint256 _maxSupply) ERC20("Credbull", "CBL") ERC20Permit("test") Ownable(_owner) {
```
Comment (pasviegas) [inline packages/contracts/src/token/CBL.sol:16]: `ERC20Permit("test") ` <- is this correct?

File: packages/contracts/src/token/CBL.sol [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    function mint(address to, uint256 amount) external onlyOwner {
+        if (totalSupply() + amount > maxSupply) {
+            revert CBL__MaxSupplyExceeded();
+        }
+        _mint(to, amount);
+    }
+
+    function burn(address from, uint256 amount) external onlyOwner {
+        _burn(from, amount);
+    }
```
Comment (lucasia) [inline packages/contracts/src/token/CBL.sol:29]: suggest we keep to the burn approach from OpenZeppelin.  this is not a restricted operation.  anyone can choose to burn their own tokens.  if we wish to restrict, we need to override update.
```
    /**
     * @dev Destroys a `value` amount of tokens from `account`, lowering the total supply.
     * Relies on the `_update` mechanism.
     *
     * Emits a {Transfer} event with `to` set to the zero address.
     *
     * NOTE: This function is not virtual, {_update} should be overridden instead
     */
    function _burn(address account, uint256 value) internal {
        if (account == address(0)) {
            revert ERC20InvalidSender(address(0));
        }
        _update(account, address(0), value);
    }
```

File: packages/contracts/src/token/CBL.sol (hunk 3 above)
Comment (lucasia) [reply packages/contracts/src/token/CBL.sol:11]: should be AccessControl.  i will update.

**Review by lucasia — APPROVED** [review]
Looks good.  We need to retrofit HelperConfig to use the Token, but that can be done on main.

File: packages/contracts/src/token/CBL.sol (hunk 4 above)
Comment (pasviegas) [reply packages/contracts/src/token/CBL.sol:33]: @lucasia @Krishnakumarskr

Comment (lucasia) [reply packages/contracts/src/token/CBL.sol:33]: no, max supply does not go down.  i've changed to be immutable to make that clear.

**Pull request merged**